src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -437,11 +437,39 @@ class MacroAssembler: public Assembler {
     Assembler::INSN(Rd, Rn, Rm, Ra);                                          \
   }
 
-  WRAP(madd) WRAP(msub) WRAP(maddw) WRAP(msubw)
+  WRAP(madd) WRAP(maddw)
   WRAP(smaddl) WRAP(smsubl) WRAP(umaddl) WRAP(umsubl)
 #undef WRAP
 
 
+  inline void msub(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_family()) {
+      /* On Neoverse, MSUB uses the same ALU with other instructions (e.g. SDIV).
+       * The combination of MUL/SUB can utilize multiple ALUs,
+       * and can be somewhat faster than MSUB. */
+      mul(rscratch1, Rn, Rm);
+      sub(Rd, Ra, rscratch1);
+    } else {
+      Assembler::msub(Rd, Rn, Rm, Ra);
+    }
+  }
+
+  inline void msubw(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_family()) {
+      /* On Neoverse, MSUB uses the same ALU with other instructions (e.g. SDIV).
+       * The combination of MUL/SUB can utilize multiple ALUs,
+       * and can be somewhat faster than MSUB. */
+      mulw(rscratch1, Rn, Rm);
+      subw(Rd, Ra, rscratch1);
+    } else {
+      Assembler::msubw(Rd, Rn, Rm, Ra);
+    }
+  }
+
```

Comment:
OK, I'm happy to approve this, but these functions are too big now to be inlined. Please put them in macroAssembler_aarch64.cpp and we're done. Thanks.

Reply:
Done in these commits:

https://github.com/openjdk/jdk/pull/19093/commits/e54be61f8c5cb6d95908e7435be5eb85b5df65fd

https://github.com/openjdk/jdk/pull/19093/commits/2a16eba516ab7ff9b90f9a2d27dc63cf5a90015b